File: ios/objc/StorylyDemo/StorylyDemo/IntegrationDemo/IntegrationViewController.m
//
//  IntegrationViewController.m
//  StorylyDemo
//
//  Created by Haldun Melih Fadillioglu on 7.02.2022.
//  Copyright © 2022 App Samurai Inc. All rights reserved.
//

#import "IntegrationViewController.h"

@interface IntegrationViewController ()

@property (weak, nonatomic) IBOutlet StorylyView *storylyView;

@end

@implementation IntegrationViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.storylyView.storylyInit = [[StorylyInit alloc] initWithStorylyId: @"eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJhY2NfaWQiOjU1NiwiYXBwX2lkIjoxMzg5LCJpbnNfaWQiOjEwNDA5fQ.kXqBdpUcKaJe7eA98PqHahMDf-123Uhb82t_mYzbBUM"];
    self.storylyView.rootViewController = self;
    self.storylyView.delegate = self;
}

- (NSString *)segueFromActionUrl:(NSString *)url {
    if (url == nil) { return nil; }
    NSURL * actionUrl = [NSURL URLWithString: url];
    if (actionUrl == nil || ![actionUrl.scheme  isEqual: @"app"] || ![actionUrl.host isEqual: @"storyly-demo"]) { return nil; }
    NSString * segueId = actionUrl.path;
    return [self canPerformSegueWithIdentifier:segueId] ? segueId : nil;
}

- (BOOL)canPerformSegueWithIdentifier:(NSString *)identifier {
    NSArray *segueTemplates = [self valueForKey:@"storyboardSegueTemplates"];
    NSArray *filteredArray = [segueTemplates filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"identifier = %@", identifier]];
    return filteredArray.count>0;
}

- (void)storylyLoaded:(StorylyView *)storylyView
       storyGroupList:(NSArray<StoryGroup *> *)storyGroupList
           dataSource:(enum StorylyDataSource)dataSource {
    NSLog(@"[storyly] IntegrationViewController:storylyLoaded - storyGroupList size {%lu)} - source {%ld)}", (unsigned long)storyGroupList.count, (long)dataSource);
}

- (void)storylyLoadFailed:(StorylyView *)storylyView
             errorMessage:(NSString *)errorMessage {
    NSLog(@"[storyly] IntegrationViewController:storylyLoadFailed - error {%@)}", errorMessage);
}

- (void)storylyActionClicked:(StorylyView *)storylyView
          rootViewController:(UIViewController *)rootViewController
                       story:(Story *)story {
    NSLog(@"[storyly] IntegrationViewController:storylyActionClicked - story action_url {%@)}", story.media.actionUrl);
    NSString *segueId = [self segueFromActionUrl:story.media.actionUrl];
    if (segueId == nil) {
        NSLog(@"[storyly]: IntegrationViewController - action_url not valid");
        return;
    }
    [storylyView dismissWithAnimated:true completion: nil];
    [self performSegueWithIdentifier:segueId sender:self];
}

@end

File: ios/swift/StorylyDemo/StorylyDemo/IntegrationDemo/IntegrationViewController.swift
//
//  IntegrationViewController.swift
//  StorylyDemo
//
//  Created by Haldun Melih Fadillioglu on 5.02.2022.
//  Copyright © 2022 App Samurai Inc. All rights reserved.
//

import UIKit
import Storyly

let STORYLY_DEMO_TOKEN = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJhY2NfaWQiOjU1NiwiYXBwX2lkIjoxMzg5LCJpbnNfaWQiOjEwNDA5fQ.kXqBdpUcKaJe7eA98PqHahMDf-123Uhb82t_mYzbBUM"

class IntegrationViewController: UIViewController {

    @IBOutlet weak var storylyView: StorylyView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        storylyView.storylyInit = StorylyInit(storylyId: STORYLY_DEMO_TOKEN)
        storylyView.rootViewController = self
        storylyView.delegate = self
    }
    
    private func segueFromActionUrl(url: String?) -> String? {
        guard let url = url,
              let actionUrl = URL(string: url) else { return nil }
        guard actionUrl.scheme == "app" && actionUrl.host == "storyly-demo" else { return nil }
        let segueId = actionUrl.path
        return isSegueFound(withIdentifier: segueId) ? segueId : nil
    }
    
    private func isSegueFound(withIdentifier id: String) -> Bool {
        guard let segues = self.value(forKey: "storyboardSegueTemplates") as? [NSObject] else { return false }
        return segues.first { $0.value(forKey: "identifier") as? String == id } != nil
    }
}

extension IntegrationViewController: StorylyDelegate {
    func storylyLoaded(_ storylyView: StorylyView, storyGroupList: [StoryGroup], dataSource: StorylyDataSource) {
        print("[storyly] IntegrationViewController:storylyLoaded - storyGroupList size {\(storyGroupList.count)} - source {\(dataSource)}")
    }
    
    func storylyLoadFailed(_ storylyView: StorylyView, errorMessage: String) {
        print("[storyly] IntegrationViewController:storylyLoadFailed - error {\(errorMessage)}")
    }
    
    func storylyActionClicked(_ storylyView: StorylyView, rootViewController: UIViewController, story: Story) {
        print("[storyly] IntegrationViewController:storylyActionClicked - story action_url {\(story.media.actionUrl ?? "")}")
        guard let segueId = segueFromActionUrl(url: story.media.actionUrl) else {
            print("[storyly] action_url not valid")
            return
        }
        storylyView.dismiss(animated: true)
        performSegue(withIdentifier: segueId, sender: self)
    }
}
